File: 0611-valid-triangle-number/0611-valid-triangle-number.java
class Solution {

    public int bs(int[] nums, int l, int r, int sum) {
        int ans = l - 1;
        while (l <= r) {
            int mid = (l + r) >> 1;

            if (nums[mid] < sum) {
                ans = mid;
                l = mid + 1;
            } else
                r = mid - 1;
        }
        return ans;
    }

    public int triangleNumber(int[] nums) {
        Arrays.sort(nums);
        int n = nums.length;
        int ans = 0;

        for (int i = 0; i < n; i++) {
            if (nums[i] == 0)
                continue;

            for (int j = i + 1; j < n; j++) {
                int sum = nums[i] + nums[j];
                int k = bs(nums, j + 1, n - 1, sum);

                ans += (k - j);
            }
        }
        return ans;
    }
}

File: 0611-valid-triangle-number/0611-valid-triangle-number.py
class Solution:
    def triangleNumber(self, nums: List[int]) -> int:
        nums.sort()
        ans, n = 0, len(nums)

        for k in range(n - 1, 1, -1):
            i, j = 0, k - 1

            # try all pairs (i, j) with this k
            while i < j:
                if nums[i] + nums[j] > nums[k]:
                    # all pairs from i..j-1 with j also valid
                    ans += (j - i)
                    j -= 1
                else:
                    i += 1

        return ans
        